Mathlib/GroupTheory/SpecificGroups/Cyclic.lean

```diff
@@ -680,3 +680,83 @@ lemma not_isCyclic_iff_exponent_eq_prime [Group α] {p : ℕ} (hp : p.Prime)
+    `addOrderOf g`. -/
+theorem zmultiplesHom_ker_eq [AddGroup G] (g : G) :
+    (zmultiplesHom G g).ker = zmultiples ↑(addOrderOf g) := by
+  ext
+  simp_rw [AddMonoidHom.mem_ker, mem_zmultiples_iff, zmultiplesHom_apply,
+    ← addOrderOf_dvd_iff_zsmul_eq_zero, zsmul_eq_mul', Int.cast_id, dvd_def, eq_comm]
+
+/-- The kernel of `zpowersHom G g` is equal to the subgroup generated by `orderOf g`. -/
+theorem zpowersHom_ker_eq [Group G] (g : G) :
+    (zpowersHom G g).ker = zpowers (Multiplicative.ofAdd ↑(orderOf g)) :=
+  congr_arg AddSubgroup.toSubgroup <| zmultiplesHom_ker_eq (Additive.ofMul g)
+
+/-- The isomorphism from `ZMod n` to any cyclic additive group of `Nat.card` equal to `n`. -/
+noncomputable def zmodAddCyclicAddEquiv [AddGroup G] (h : IsAddCyclic G) :
+    ZMod (Nat.card G) ≃+ G := by
+  let n := Nat.card G
+  let ⟨g, surj⟩ := Classical.indefiniteDescription _ h.exists_generator
+  have kereq : ((zmultiplesHom G) g).ker = zmultiples ↑(Nat.card G) := by
+    rw [zmultiplesHom_ker_eq]
+    congr
+    rw [← Nat.card_zmultiples]
+    exact Nat.card_congr (Equiv.subtypeUnivEquiv surj)
+  exact (Int.quotientZmultiplesNatEquivZMod n)
+    |>.symm.trans <| QuotientAddGroup.quotientAddEquivOfEq kereq
+    |>.symm.trans <| QuotientAddGroup.quotientKerEquivOfSurjective (zmultiplesHom G g) surj
+
+/-- The isomorphism from `Multiplicative (ZMod n)` to any cyclic group of `Nat.card` equal to `n`.
+-/
+noncomputable def zmodCyclicMulEquiv [Group G] (h : IsCyclic G) :
+    Multiplicative (ZMod (Nat.card G)) ≃* G :=
+  AddEquiv.toMultiplicative <| zmodAddCyclicAddEquiv <| isAddCyclic_additive_iff.2 h
+
+/-- Two cyclic additive groups of the same cardinality are isomorphic. -/
+noncomputable def addEquivofAddCyclicCardEq [AddGroup G] [AddGroup H] [hG : IsAddCyclic G]
+    [hH : IsAddCyclic H] (hcard : Nat.card G = Nat.card H) : G ≃+ H := hcard ▸
+  zmodAddCyclicAddEquiv hG |>.symm.trans (zmodAddCyclicAddEquiv hH)
+
+/-- Two cyclic groups of the same cardinality are isomorphic. -/
+@[to_additive existing addEquivofAddCyclicCardEq]
+noncomputable def mulEquivOfCyclicCardEq [Group G] [Group H] [hG : IsCyclic G]
+    [hH : IsCyclic H] (hcard : Nat.card G = Nat.card H) : G ≃* H := hcard ▸
+  zmodCyclicMulEquiv hG |>.symm.trans (zmodCyclicMulEquiv hH)
+
+/-- Two groups of the same prime cardinality are isomorphic. -/
+@[to_additive "Two additive groups of the same prime cardinality are isomorphic."]
+noncomputable def mulEquivOfPrimeCardEq {p : ℕ} [Fintype G] [Fintype H] [Group G] [Group H]
+    [Fact p.Prime] (hG : Fintype.card G = p) (hH : Fintype.card H = p) : G ≃* H := by
+  have hGcyc := isCyclic_of_prime_card hG
+  have hHcyc := isCyclic_of_prime_card hH
+  apply mulEquivOfCyclicCardEq
+  rw [← Nat.card_eq_fintype_card] at hG hH
+  exact hG.trans hH.symm
+
+attribute [to_additive existing zmultiplesHom_ker_eq] zpowersHom_ker_eq
+attribute [to_additive existing] zmodCyclicMulEquiv
+
```

Comment:
I don't think these are actually supported by `toAdditive`, as their statements are not similar enough (one has `Multiplicative`). I'd suggest removing them.

Reply:
All of them or just these two?

Reply:
Oh I didn't see your earlier commit, so "all of them" and "just these two" are now synonymous.

Reply:
So `zpowersHom` and `zadditiveHom` involve `Multiplicative` and use `@[to_additive existing ...]`, which was my reason to add these.

Reply:
I see.